### PR TITLE
util/types: workaround for time roundFrac

### PR DESCRIPTION
**Pull request by tiancaiamao**
This is a workaround for `roundFrac` function, take time part out and do rounding, so we can handle month or day contains 0.

`YearDay` is also updated use the new time implementation.

After this PR, nearly all `TODO: Fix here.` are removed, means it's safe to use datetime month or day contains 0.

### CODE REVIEW

File: server/util.go [hunk 1]
```diff
@@ -195,10 +195,7 @@ func dumpBinaryTime(dur time.Duration) (data []byte) {
 
 func dumpBinaryDateTime(t types.Time, loc *time.Location) (data []byte) {
 	if t.Type == mysql.TypeTimestamp && loc != nil {
-		t1, err := t.Time.GoTime()
-		if err != nil {
-			// TODO: Fix here.
-		}
+		t1, _ := t.Time.GoTime()
```
Comment (zimulala) [inline server/util.go]: Why remove the comments here?

Comment (tiancaiamao) [inline server/util.go]: because it's safe to convert `mysql.TypeTimestamp` to go Time, it won't return error.

**Comment by tiancaiamao** [comment]
PTAL @shenli @zimulala

File: util/types/time_test.go [hunk 2]
```diff
@@ -587,6 +587,9 @@ func (s *testTimeSuite) TestRoundFrac(c *C) {
 		{"2012-12-31 11:30:45.123456", 1, "2012-12-31 11:30:45.1"},
 		{"2012-12-31 11:30:45.999999", 4, "2012-12-31 11:30:46.0000"},
 		{"2012-12-31 11:30:45.999999", 0, "2012-12-31 11:30:46"},
+		{"2012-00-00 11:30:45.999999", 3, "2012-00-00 11:30:46.000"},
+		// MySQL can handle this case, but we can't.
```
Comment (hanfei1991) [inline util/types/time_test.go]: add todo?

File: util/types/time.go [hunk 3]
```diff
@@ -327,10 +327,26 @@ func (t Time) roundFrac(fsp int) (Time, error) {
 		return t, nil
 	}
 
-	t1, _ := t.Time.GoTime()
-	// TODO: Fix here.
-	nt := t1.Round(gotime.Duration(math.Pow10(9-fsp)) * gotime.Nanosecond)
-	return Time{Time: FromGoTime(nt), Type: t.Type, Fsp: fsp}, nil
+	var nt TimeInternal
+	if t1, err := t.Time.GoTime(); err == nil {
```
Comment (hanfei1991) [inline util/types/time.go]: Do we need check the error type?

File: util/types/time.go [hunk 4]
```diff
@@ -327,10 +327,26 @@ func (t Time) roundFrac(fsp int) (Time, error) {
...
+		nt = FromGoTime(t1.Round(gotime.Duration(math.Pow10(9-fsp)) * gotime.Nanosecond))
```
Comment (hanfei1991) [inline util/types/time.go]: abstract t1.Round(gotime.Duration(math.Pow10(9-fsp)) * gotime.Nanosecond) as some thing?

File: util/types/time.go [hunk 5]
```diff
@@ -327,10 +327,26 @@ func (t Time) roundFrac(fsp int) (Time, error) {
...
+		// Take the hh:mm:ss part out to avoid handle month or day = 0.
+		hour, minute, second, microsecond := t.Time.Hour(), t.Time.Minute(), t.Time.Second(), t.Time.Microsecond()
+		t1 := gotime.Date(1, 1, 1, hour, minute, second, microsecond*1000, gotime.Local)
+		t2 := t1.Round(gotime.Duration(math.Pow10(9-fsp)) * gotime.Nanosecond)
+		hour, minute, second = t2.Clock()
+		microsecond = t2.Nanosecond() / 1000
+		day := t2.Day() - 1
+
+		// TODO: Can't handle this currently, but never mind.
+		if day > 0 {
```
Comment (hanfei1991) [inline util/types/time.go]: When the day will be greater than 0? add comment.

File: util/types/time_test.go (hunk 2 above)
Comment (tiancaiamao) [inline util/types/time_test.go]: This is not a big problem, it must satisfy conditions:
1. time round overflow one day, so it must be 23.59.59.9999999
2. date part invalid, contains month = 0 or day =0

The case is so rare that may be it won't be done in the future ...

File: util/types/time.go (hunk 3 above)
Comment (tiancaiamao) [inline util/types/time.go]: error handle goes to another branch, see line 334

File: util/types/time.go (hunk 4 above)
Comment (tiancaiamao) [inline util/types/time.go]: it's too many arguments write in one line, I'll clean it.

File: util/types/time.go [hunk 6]
```diff
@@ -327,10 +333,28 @@ func (t Time) roundFrac(fsp int) (Time, error) {
...
+		hour, minute, second, microsecond := t.Time.Hour(), t.Time.Minute(), t.Time.Second(), t.Time.Microsecond()
+		t1 := gotime.Date(1, 1, 1, hour, minute, second, microsecond*1000, gotime.Local)
+		t2 := roundTime(t1, fsp)
+		hour, minute, second = t2.Clock()
+		microsecond = t2.Nanosecond() / 1000
+		day := t2.Day() - 1
+
+		// TODO: day > 0 means hh:mm:ss overflow one day after rounding, it should be add to yy:mm:dd part,
+		// but mm:dd may contain 0, it makes the code complex, so we ignore it here.
+		if day > 0 {
```
Comment (zimulala) [inline util/types/time.go]: Use `t2.Day()>1` directly.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**